common/src/main/java/org/corfudb/common/util/CompatibilityVectorUtils.java

```diff
@@ -0,0 +1,118 @@
+package org.corfudb.common.util;
+
+import com.google.common.annotations.VisibleForTesting;
+import lombok.extern.slf4j.Slf4j;
+import org.corfudb.common.util.proto.CorfuFeatures.Feature;
+
+import java.util.BitSet;
+
+/**
+ * A collection of {@link BitSet}s representing the Corfu features enabled so far.
+ *
+ * Creates the Compatibility Vectors that contain the information about the Corfu features enabled
+ * so far as per the current stage of the code during the compile time of the class.
+ * The vectors are of type {@link BitSet} in which each bit will represent if the corresponding
+ * feature is enabled or not based on its value (0 or 1).
+ *
+ * Client-Server Transmission:
+ * These BitSet values will be converted to array of bytes (byte[]) set in the Protobuf header
+ * HeaderMsg.ProtocolVersionMsg value (of both RequestMsg and ResponseMsg),
+ * and sent to the receiver. The receiver will then convert the array of bytes into BitSet and check
+ * the specific features using
+ * {@link CompatibilityVectorUtils#isFeatureEnabled(Feature, BitSet)} method.
+ *
+ * Steps for New Features:
+ * 1. <u>Introduce</u> them by creating an entry in the file corfu_features.proto
+ * 2. <u>Enable</u> them by appending to values of the field - 'supportedFeatures' in this class.
+ *
+ * Created by cgudisagar on 5/3/21.
+ */
+@Slf4j
+public class CompatibilityVectorUtils {
+
+    /**
+     * This initializer class loads when getCompatibilityVectors() is called for the first time.
+     * It gives a thread-safe lazy-initialization because the class loader guarantees that all
+     * static initialization is complete before getting access to the class.
+     */
+    @VisibleForTesting
+    private static class CompatibilityVectorWrapper {
+
+        /**
+         * Contains the vectors representing the Corfu features enabled so far.
+         */
+        private static final BitSet vectors = generateVectors();
+
+        private static BitSet generateVectors(){
+            // length-1 to get rid of the default UNRECOGNIZED enum value, else we get an error
+            BitSet bitSet = new BitSet(Feature.values().length-1);
+            Feature[] features = Feature.values();
+            for(int feature_index=1; feature_index < features.length-1; ++feature_index){
```

Comment:
Should this be 
for (int feature_index=1; feature_index < features.length; ++feature_index) {
Since you already have a < check should it less than length instead of length - 1 ?

Reply:
Sorry for the confusion, Hope this helps -

For a given enum - 
```
enum Feature {
  FEATURE_INVALID = 0;
  TEST = 1;
  ABC = 2;
}
```

Feature.values() returns `["FEATURE_INVALID", "TEST", "ABC", "UNRECOGNIZED"]` (a list of enum values, not exactly strings)
Note: `length = 4`

We want the result bitset to contain the first three elements i.e. length-1 elements.

The `for` loop **should** start from index 1, else it will set the INVALID feature to 1. We want it to be 0 as it's an INVALID feature by convention.

Further, the loop can end at index 2 (`< length -1`), otherwise, it will try to access the UNRECOGNIZED feature at index 3, which gives an error at runtime. I had added the `if` checks for safety. I will remove them.